### ISSUE TITLE
DietPi-Config | Allow to set static DNS with DHCP

### ISSUE THREAD

**Issue by krishnasen2401** [MichaIng/DietPi#6423]
#### Is your feature request related to a problem? Please describe:

- ...
Some times I keep changing router when using ethernet connection some have different gateway address its always a issue to set the ip to static and back when changing in between just to use static dns
#### Describe the solution you'd like:

- ...
a Static dns option like the ones we in android, windows and Mac so that we can DNS without changing the IP as well
#### Describe alternatives you've considered:

- ...
can be changed to using cmd but it always creates problems
#### Additional context

- ...

**Comment by JappeHallunken** [comment]
I think the problem here is, whenever you config your network with `dietpi-config`, your settings are written to `/etc/network/interfaces` and, if available, the DNS settings are written to `/etc/resolv.conf`.
You would need to avoid using `dietpi-config` and edit these two files directly. If you use DHCP you would only need to change your DNS settings in `/etc/resolv.conf`.

**Comment by MichaIng** [comment]
I'm not sure whether this is natively possible with `ifupdown` resp. `/etc/network/interfaces`:
- If one configures an interface with DHCP, the DHCP client sets everything, the IP address as well as the DNS resolver, if provided by the DHCP server.
- If one sets an interface to "static", everything is set statically once.

Searching a bit around, this can be tested: https://askubuntu.com/a/623726
```sh
G_CONFIG_INJECT 'supersede[[:blank:]]+domain-name-servers[[:blank:]]' 'supersede domain-name-servers 1.1.1.1;' /etc/dhcp/dhclient.conf
```

Jep, that works here, just tested in a VM.

**Comment by potassiumchloride** [comment]
I would like to chime in and ask if this feature request could be implemented in the near future. In my opinion it is a helpful feature to give the option to override a DHCP assigned DNS server (from the router) by a static DNS entry via `dietpi-config`.

Background:
My router works as DHCP server and assigns IP addresses to the clients and hands out the DNS server (which is a Pi-hole on a Raspberry Pi running dietpi). I prefer to not set static IP addresses on any client but use DHCP only, as this allows me to easily (physically) transport clients to other networks from time to time (even the Pi-hole). However, all clients permanently receive the same IP address by "static DHCP" assignment, i. e. the Pi-hole too. Consequently, this means that the Pi-hole computer also gets its own IP address as DNS server by the router's DHCP. This is considered bad practice, for the Pi-hole's underlying operating system should never use the Pi-hole itself as DNS server; for example it could result in blocking system updates if the Pi-hole is down/broken.

Hence, I manually override in dietpi the DHCP assigned DNS server in `/etc/resolv.conf` by editing `/etc/dhcpcd.conf` [before it was `/etc/dhcp/dhclient.conf` (see [here](https://github.com/MichaIng/DietPi/issues/6423#issuecomment-1587923206)), but since dietpi's recent changes it is `/etc/dhcpcd.conf` now]:

`# A list of options to request from the DHCP server.`
`# *** Override default DNS from DHCP server and set static DNS`
`# entries (the router serves as DNS server) for the Pi-hole OS ***`
`#option domain_name_servers, domain_name, domain_search`
`option domain_name, domain_search`
`static domain_name_servers=IPv4_address_of_router IPv6_of_router`

Since these manual edits could be overridden by future updates of the `dhcpcd-base` package, it would be a great option to set it permanently via `dietpi-config` as suggested [here](https://github.com/MichaIng/DietPi/issues/6423#issuecomment-1587923206).

EDIT: Some typo changes

**Comment by MichaIng** [comment]
So your Pi-hole host is requesting its own IP address from its own Pi-hole? I understand the benefit of reserved DHCP addresses vs static IP, but I wouldn't do that for the DHCP server host itself. It could break its own network access, hence the whole LAN, during maintenance or failures. Basically the same reason why you do not want to use Pi-hole for its own DNS.

We btw reverted to `dhclient`, at last on Bookworm and Trixie systems (Forky has no `dhclient` anymore). And I plan to move to `udhcpc` once the bugs in Debian's `ifupdown` are solved. `udhcpc` has no config file, which makes an implementation at the DHCP client side difficult.

So we'll probably need to look into `ifupdown` resp. `/etc/network/interfaces`, or use `resolvconf`, e.g. via `/etc/resolvconf/resolv.conf.d/head` or `/etc/resolvconf/resolv.conf.d/base`. However, AFAIK this can only add a DNS resolver, not remove/replace one added by the DHCP client.

**Comment by potassiumchloride** [comment]
Thanks for your immediate response; let me give some details in a more concrete manner because I think you mixed it up a bit (probably I did not explain it detailed enough):

> So your Pi-hole host is requesting its own IP address from its own Pi-hole? I understand the benefit of reserved DHCP addresses vs static IP

No, the Pi-hole host (= a Raspi running dietpi) is not requesting its own IP addresse from Pi-hole, The DHCP server functionality on Pi-hole is of course deactivated (default Pi-hole setting). The Pi-hole host requests its own IP address from the router (a Fritzbox acting as DHCP server and handing out only reserved IP address assignments). The same router's DHCP server assigns the IPv4 and IPv6 address of the Pi-hole as DNS server to all clients, i. e. the Pihole host system receives its own IPv4/v6 addresses via DHCP and stores them in `/etc/resolv.conf` (which is bad practice for the reasons mentioned).

> but I wouldn't do that for the DHCP server host itself. It could break its own network access, 

Exactly. However, it is not the DHCP server host itself in my case, but the host OS (dietpi) which runs the local DNS server for the entire home network (the Pi-Hole on top of dietpi).

The whole issue of potentially blocking its own network access I try to circumvent by changing the stored entries in `/etc/resolv.conf` (obtained from DHCP on dietpi) via overriding the settings in `/etc/dhcpcd.conf` (see previous comment).

Maybe the use case is too specialized to implement as a general option in dietpi (its default should be "off" anyway for the general user), especially considering the issues you might face after I read your additional remarks with respect to future DHCP implementation switches in dietpi. But since also another used had already asked for this feature request, I thought I should join to express my wish for this addition. :-)

**Comment by Joulinar** [comment]
Simply switch to Static IP on your PiHole or switch DHCP functionality to PiHole.

**Comment by MichaIng** [comment]
> The DHCP server functionality on Pi-hole is of course deactivated

Ah okay, that way round. One could still argue that the local DNS server, same as the local DHCP server, both should better have a static IP. But I get the mobility aspect.

Anyway, I agree that feature would be handy. But I myself will rather work on #6566 first, splitting all the network setup reworked into a new script, to have a better basis for new network features.

**Comment by potassiumchloride** [comment]
> But I get the mobility aspect. Anyway, I agree that feature would be handy.

Yes, that's my main reason why I have not set up static IP addresses. Thanks for understanding my point.
 
>  But I myself will rather work on [#6566](https://github.com/MichaIng/DietPi/issues/6566) first, splitting all the network setup reworked into a new script, to have a better basis for new network features.

Yes, no problem! I just wanted to let the development team know that the implementation of this feature could be wanted by other people as well. If you are ready to tackle this issue and find time to work on it, I will be happy to help by testing it.

**Comment by potassiumchloride** [comment]
> Simply switch to Static IP on your PiHole or switch DHCP functionality to PiHole.

If I had wanted to do any of that, I would already have done so long ago and not tried to support this feature request of the original poster here and now. If the development team wants to add this feature, I will be happy to help and test it.

**Comment by Joulinar** [comment]
Don't expect this to happen any time soon.

**Comment by potassiumchloride** [comment]
> Don't expect this to happen any time soon.

No problem; patience is a virtue and I already live with my workaround of manually overriding settings in `/etc/dhcpcd.conf`.
